### PR TITLE
Remove decommission from RestorationBuckets

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

Removes the concept of decommissioning a bucket from the Restoration Framework because it didn't work in all cases (see "widgets moves scopes during restore" test case). Basically, when doing a restore turns on restoration for a widget that previously didn't have it turned on, that widget wouldn't properly restore its data. The reason: Since restoration was turned off for the widget, it didn't have a bucket to listen to the decommission notification. Therefore, when it got a bucket during the restore, it thought that it had just moved into a restoration scope and it persisted its existing state in the bucket instead of restoring itself.

The problem has been fixed by adding a `isReplacing` getter to the bucket. If a widget receives a bucket and that flag is set, it knows that it is supposed to restore its state from the information in the bucket instead of persisting its current state into it.

In addition to this, a new `initialRestore` flag is added to `RestorationMixin.restoreState`. It can be used by widgets to determine whether they should use the current restore to initialize their state or whether this is a subsequent restore and they need to just update their state.

## Related Issues

n/a

## Tests

Yes

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by gspencergoog** [comment]
Looks like the analyzer isn't happy yet.

File: packages/flutter/test/widgets/restoration_mixin_test.dart [hunk 1]
```diff
@@ -127,6 +127,7 @@ void main() {
 
     // Rename the existing bucket.
     state.injectId('newnewnew');
+    await tester.pump();
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/restoration_mixin_test.dart:130]: Why the extra pump?

File: packages/flutter/test/widgets/restoration_mixin_test.dart [hunk 2]
```diff
@@ -721,6 +678,7 @@ class _TestRestorableWidgetState extends State<_TestRestorableWidget> with Resto
 
   @override
   void didToggleBucket(RestorationBucket oldBucket) {
+    print('$oldBucket -> $bucket');
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/restoration_mixin_test.dart:681]: Remove debug print?

File: packages/flutter/test/widgets/restoration_scopes_moving_test.dart [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  testWidgets('widgets moves scopes during restore', (WidgetTester tester) async {
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/restoration_scopes_moving_test.dart:11]:
```suggestion
  testWidgets('widgets move scopes during restore', (WidgetTester tester) async {
```

File: packages/flutter/test/widgets/root_restoration_scope_test.dart [hunk 4]
```diff
@@ -362,7 +362,7 @@ void main() {
     expect(state.bucket, isNotNull);
 
     binding.restorationManager.rootBucket = SynchronousFuture<RestorationBucket>(null);
-    root..decommission()..dispose();
+    root.dispose();
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/root_restoration_scope_test.dart:365]: Why did the one above move to after the pump, and this one didn't?

Comment (goderbauer) [reply packages/flutter/test/widgets/root_restoration_scope_test.dart:365]: Moved both to after.

File: packages/flutter/test/widgets/restoration_mixin_test.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/restoration_mixin_test.dart:130]: Removed. It is not necessary. Must be left over from debugging something.

**Comment by goderbauer** [comment]
All comments addressed. PTAL @gspencergoog

File: packages/flutter/lib/src/services/restoration.dart [hunk 5]
```diff
@@ -529,6 +524,16 @@ class RestorationBucket extends ChangeNotifier {
   RestorationManager? _manager;
   RestorationBucket? _parent;
 
+  /// Returns true when entities processing this bucket should restore their
+  /// state from the information in the bucket (e.g. via [read] and
+  /// [claimChild]) instead of copying their current state information into the
+  /// bucket (e.g. via [write] and [adoptChild].
+  ///
+  /// This flag is true for one frame after the [RestorationManager] has been
```
Comment (gspencergoog) [inline packages/flutter/lib/src/services/restoration.dart:532]: nit: It's unlikely, but the "is true for one frame after" could be misconstrued as "is true for one of the frames after" (i.e. not just the immediately following frame, but one of any of the following frames).

Maybe say "is true for the frame after" instead? Here and above.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by devoncarew** [comment]
@goderbauer - I believe this PR is causing the flutter-analyze try bot (a Dart chromium bot) to fail:

https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8872044253012641440/+/steps/analyze_flutter/0/stdout

That bot validates the version of the analysis server at head for dart-lang/sdk against the current code in flutter/flutter.

The issue is:

```
warning • The parameter 'data' is required • packages/flutter_test/lib/src/restoration.dart:63:5 • missing_required_param
```

**Comment by goderbauer** [comment]
https://github.com/flutter/flutter/pull/63728 is the fix.

**Pull request merged**